### PR TITLE
Feature/create reporting year

### PR DESCRIPTION
**Pull request by blueQcloud**

### CODE REVIEW

File: schema/deploy/table_application.sql [hunk 1]
```diff
@@ -4,12 +4,14 @@
...
 -- requires: table_form_result
+-- requires: table_reporting_year
 
 begin;
 
 create table ggircs_portal.application (
     id integer primary key generated always as identity,
-    facility_id integer not null references ggircs_portal.facility(id)
+    facility_id integer not null references ggircs_portal.facility(id),
+    foreign key (reporting_year) references ggircs_portal.reporting_year(reporting_year)
```
Comment (matthieu-foucault) [inline schema/deploy/table_application.sql]:
```suggestion
   reporting_year integer not null references ggircs_portal.reporting_year(reporting_year)
```

Comment (matthieu-foucault) [reply schema/deploy/table_application.sql]: You will need to add a comment for the column as our style tests require it

File: schema/deploy/table_reporting_year.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+begin;
+
+create table ggircs_portal.reporting_year (
+    reporting_year integer primary key,
+    reporting_period_start timestamp with time zone,
+    reporting_period_end timestamp with time zone,
+    application_open_date timestamp with time zone,
+    application_end_date timestamp with time zone,
+    application_response_date timestamp with time zone
```
Comment (matthieu-foucault) [inline schema/deploy/table_reporting_year.sql]: All of those need some `not null` constraint

File: schema/deploy/table_reporting_year.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+);
+
+commit;
```
Comment (matthieu-foucault) [inline schema/deploy/table_reporting_year.sql]: You need to add comments to the table and to each column

File: schema/sqitch.plan [hunk 4]
```diff
@@ -48,4 +49,4 @@ function_init_application_administration_form_result [table_form_result function
 function_init_application_fuel_form_result [table_form_result function_get_swrs_fuel_data] 2019-11-12T18:00:52Z Matthieu Foucault <matthieu@button.is> # add function to initialize fuel form
 function_init_application_emission_form_result [table_form_result function_get_swrs_emission_data] 2019-11-12T18:01:51Z Matthieu Foucault <matthieu@button.is> # add function to initialize emission form
 function_facility_has_swrs_report [table_facility] 2019-11-12T19:25:09Z Matthieu Foucault <matthieu@button.is> # add function to check if a facility has a swrs report
-function_search_ciip_user_organisation [table_ciip_user_organisation] 2019-11-06T18:34:19Z Helen Liu <helen@button.is> # Added search function on user organisation table
+function_search_ciip_user_organisation [table_ciip_user_organisation] 2019-11-06T18:34:19Z Helen Liu <helen@button.is> # Added search function on user organisation table\n:
```
Comment (matthieu-foucault) [inline schema/sqitch.plan]:
```suggestion
function_search_ciip_user_organisation [table_ciip_user_organisation] 2019-11-06T18:34:19Z Helen Liu <helen@button.is> # Added search function on user organisation table
```

File: schema/data/portal/reporting_year.sql [hunk 5]
```diff
@@ -0,0 +1,14 @@
+begin;
+
+with rows as (
+insert into ggircs_portal.reporting_year(reporting_year, reporting_period_start, reporting_period_end, application_open_date, application_end_date, application_response_date)
+overriding system value
+values
+(2018, '2017-01-01 14:49:54.191757-07', '2018-12-30 14:49:54.191757-07', '2018-04-01 14:49:54.191757-07', '2018-12-30 14:49:54.191757-07', '2019-01-30 14:49:54.191757-07'),
+(2019, '2018-01-01 14:49:54.191757-07', '2019-12-30 14:49:54.191757-07', '2019-04-01 14:49:54.191757-07', '2019-12-30 14:49:54.191757-07', '2020-01-30 14:49:54.191757-07')
+on conflict(reporting_year) do update set
+reporting_year=excluded.reporting_year
```
Comment (matthieu-foucault) [inline schema/data/portal/reporting_year.sql]: in the list of columns to update on conflict, you need to have everything **but** the column that is conflicting.

What you wrote is telling the database: "if there is already a row with the `reporting_year` 2018, set the reporting year to 2018". 
What you want is "if there is already a row with the `reporting_year` 2018, set `reporting_period_start` to the value we tried to insert (`excluded.reporting_period_start`), etc for every column other than the primary key.

File: app/pages/index.tsx [hunk 6]
```diff
@@ -17,13 +19,23 @@ export default class Index extends Component<Props> {
...
   render() {
     const {query} = this.props;
     const {session} = query || {};
+    const date1: any = query.getReportingYear.applicationOpenDate;
+    const startDate: Date = new Date(Date.parse(date1));
+    const date2: any = query.getReportingYear.applicationEndDate;
+    const endDate: Date = new Date(Date.parse(date2));
+    const time1 = startDate.toDateString();
+    const time2 = endDate.toDateString();
     return (
```
Comment (dleard) [inline app/pages/index.tsx]: time1 & time2 are not very descriptive variable names. Something along the lines of parsedStartDate & parsedEndDate better describe the variables' purpose.

File: app/pages/index.tsx [hunk 7]
```diff
@@ -8,6 +8,8 @@ import RegistrationLoginButtons from '../components/RegistrationLoginButtons';
 
 interface Props extends CiipPageComponentProps {
   query: pagesQueryResponse['query'];
+  applicationOpenDate?: string;
+  applicationEndDate?: string;
```
Comment (matthieu-foucault) [inline app/pages/index.tsx]: Those props seem to be incorrect

File: app/tests/pages/__snapshots__/index.test.tsx.snap [hunk 8]
```diff
@@ -188,7 +188,7 @@ exports[`landing It matches the last accepted Snapshot 1`] = `
             <td
               className="jsx-473442302"
             >
-              May 1, 2020
+              Mon Apr 01 2019
```
Comment (matthieu-foucault) [inline app/tests/pages/__snapshots__/index.test.tsx.snap]: Not sure about this format, can you change `toDateString` to output in the same format as the hardcoded values we had before?

File: schema/deploy/function_get_reporting_year.sql [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    as $function$
+
+            select *
+            from ggircs_portal.reporting_year
+            where (select CURRENT_TIMESTAMP) between reporting_period_start::timestamp with time zone and reporting_period_end::timestamp with time zone
+
+    $function$
+        language sql stable;
+
+commit;
```
Comment (matthieu-foucault) [inline schema/deploy/function_get_reporting_year.sql]: add a comment to the function

**Comment by matthieu-foucault** [comment]
@dleard can you give me a chance to review the PR before dismissing my review?

File: schema/deploy/function_get_reporting_year.sql [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+begin;
+
+create or replace function ggircs_portal.get_reporting_year()
+    returns ggircs_portal.reporting_year
+    as $function$
+
+            select *
+            from ggircs_portal.reporting_year
+            where (select CURRENT_TIMESTAMP) between reporting_period_start::timestamp with time zone and reporting_period_end::timestamp with time zone
```
Comment (matthieu-foucault) [inline schema/deploy/function_get_reporting_year.sql]: This seems incorrect. We want to return the reporting year 2019 if the current timestamp is in 2019, but `reporting_period_start` and `reporting_period_end` are in 2018 for that year

**Comment by matthieu-foucault** [comment]
@blueQcloud please look at the outstanding comments above and open another PR to fix those

**Pull request merged**